server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/bitbuy/BitbuyExchange.java

```diff
@@ -0,0 +1,329 @@
+        return call("send coins", () -> api.withdraw(cryptoCurrency, destinationAddress, amount).transactionReference);
+    }
+
+    @Override
+    public ITask createPurchaseCoinsTask(BigDecimal amount, String cryptoCurrency, String fiatCurrencyToUse, String description) {
+        return createOrderTask(amount, cryptoCurrency, fiatCurrencyToUse, OrderSide.BUY);
+    }
+
+    @Override
+    public ITask createSellCoinsTask(BigDecimal amount, String cryptoCurrency, String fiatCurrencyToUse, String description) {
+        return createOrderTask(amount, cryptoCurrency, fiatCurrencyToUse, OrderSide.SELL);
+    }
+
+    private ITask createOrderTask(BigDecimal amount, String cryptoCurrency, String fiatCurrencyToUse, OrderSide orderSide) {
+        if (cryptoCurrency == null || fiatCurrencyToUse == null) {
+            return null;
+        }
+        if (!isCryptoCurrencySupported(cryptoCurrency)) {
+            return null;
+        }
+        if (!isFiatCurrencySupported(fiatCurrencyToUse)) {
+            return null;
+        }
+        return new OrderTask(amount, cryptoCurrency, fiatCurrencyToUse, orderSide);
+    }
+
+    private BigDecimal getRateSourceCryptoVolume(String cryptoCurrency) {
+        if (CryptoCurrency.BTC.getCode().equals(cryptoCurrency)) {
+            return BigDecimal.ONE;
+        }
+        return BigDecimal.TEN;
+    }
+
+    @Override
+    public BigDecimal getExchangeRateForBuy(String cryptoCurrency, String fiatCurrency) {
+        BigDecimal rateSourceCryptoVolume = getRateSourceCryptoVolume(cryptoCurrency);
+        BigDecimal result = calculateBuyPrice(cryptoCurrency, fiatCurrency, rateSourceCryptoVolume);
+        if (result != null) {
+            return result.divide(rateSourceCryptoVolume, 2, BigDecimal.ROUND_UP);
+        }
+        return null;
+    }
+
+    @Override
+    public BigDecimal getExchangeRateForSell(String cryptoCurrency, String fiatCurrency) {
+        BigDecimal rateSourceCryptoVolume = getRateSourceCryptoVolume(cryptoCurrency);
+        BigDecimal result = calculateSellPrice(cryptoCurrency, fiatCurrency, rateSourceCryptoVolume);
+        if (result != null) {
+            return result.divide(rateSourceCryptoVolume, 2, BigDecimal.ROUND_DOWN);
+        }
+        return null;
+    }
+
+    @Override
+    public BigDecimal calculateBuyPrice(String cryptoCurrency, String fiatCurrency, BigDecimal cryptoAmount) {
+        return calculatePrice(cryptoCurrency, fiatCurrency, cryptoAmount, OrderSide.BUY);
+    }
+
+    @Override
+    public BigDecimal calculateSellPrice(String cryptoCurrency, String fiatCurrency, BigDecimal cryptoAmount) {
+        return calculatePrice(cryptoCurrency, fiatCurrency, cryptoAmount, OrderSide.SELL);
+    }
+
+    private BigDecimal calculatePrice(String cryptoCurrency, String fiatCurrency, BigDecimal cryptoAmount, OrderSide orderSide) {
+        if (cryptoCurrency == null || fiatCurrency == null) {
```

Comment:
This is the same code as the one used at `createOrderTask` r. 146-154. This could be moved to some validation method

Reply:
I simplified the check so it's just one line now